test/extended/baremetal/helper.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// getExtraWorkerSecretData gets and decodes the secret associated for the
+// specified extra worker. If not found, the test will fail
+func (b *BaremetalTestHelper) getExtraWorkerSecretData(index int) *v1.Secret {
+	ew, err := b.extraWorkersRetrieveData()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	yamlData, ok := ew.Data[b.extraWorkersSecretKey(index)]
+	o.Expect(ok).To(o.BeTrue(), fmt.Sprintf("unable to find data for extra worker %d", index))
+
+	var secret v1.Secret
+	err = yaml.Unmarshal(yamlData, &secret)
+	o.Expect(err).ToNot(o.HaveOccurred())
+	return &secret
+}
+
+// getExtraWorkerSecretData gets and decodes the baremetal host associated for the
+// specified extra worker. If not found, the test will fail
+func (b *BaremetalTestHelper) getExtraWorkerData(index int) *unstructured.Unstructured {
+	ew, err := b.extraWorkersRetrieveData()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	yamlData, ok := ew.Data[b.extraWorkerKey(index)]
+	o.Expect(ok).To(o.BeTrue(), fmt.Sprintf("unable to find data for extra worker %d", index))
+
+	jsonData, err := yaml.YAMLToJSON(yamlData)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	var host unstructured.Unstructured
+	err = host.UnmarshalJSON(jsonData)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	return &host
+}
+
+// GetExtraWorkerData gets baremetal host and secret for the specified extra worker.
+// If not found, the test will fail
+func (b *BaremetalTestHelper) GetExtraWorkerData(index int) (*unstructured.Unstructured, *v1.Secret) {
+	return b.getExtraWorkerData(index), b.getExtraWorkerSecretData(index)
+}
+
+// CreateExtraWorker creates a new BaremetalHost (and associated Secret).
+// If successfull, returns the newly created host and secret resources
+func (b *BaremetalTestHelper) CreateExtraWorker(host *unstructured.Unstructured, secret *v1.Secret) (*unstructured.Unstructured, *v1.Secret) {
+
+	secret, err := b.clientSet.CoreV1().Secrets("openshift-machine-api").Create(context.Background(), secret, metav1.CreateOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	host, err = b.bmcClient.Create(context.Background(), host, metav1.CreateOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	b.extraWorkers = append(b.extraWorkers, host)
+
+	return host, secret
+}
+
+// DeleteAllExtraWorkers deletes all the extra workers created in the current session
+func (b *BaremetalTestHelper) DeleteAllExtraWorkers() {
+	for _, worker := range b.extraWorkers {
+		err := b.bmcClient.Delete(context.Background(), worker.GetName(), metav1.DeleteOptions{})
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		b.waitForDeletion(worker)
+	}
+}
+
+func (b *BaremetalTestHelper) waitForDeletion(host *unstructured.Unstructured) {
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (done bool, err error) {
+		_, err = b.bmcClient.Get(context.TODO(), host.GetName(), metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return true, nil
+			}
+
+			return false, err
```

Comment:
Just curious, this will retry every second, so if there are any problems accessing the ApiServer on any of the retries the test will fail. It sounds like this will be fragile, but I guess this is how other tests work?

Reply:
In theory it's right, but  if we have issues in the apiserver in a serial test (so nothing else is running) then the situation is probably even worse :). Another strategy could be to filter other kinds of errors, it's probably something we could tweak in future in case of need (other tests adopted also the approach to watch for a particular event, but this one seems better to me)